crates/oxc_linter/src/service/runtime.rs

```diff
@@ -2,43 +2,121 @@ use std::{
     loader::{JavaScriptSource, LINT_PARTIAL_LOADER_EXT, PartialLoader},
     module_record::ModuleRecord,
     utils::read_to_string,
 };
 
-use super::{
-    LintServiceOptions,
-    module_cache::{ModuleCache, ModuleState},
-};
-
 pub struct Runtime {
     cwd: Box<Path>,
     /// All paths to lint
-    paths: FxHashSet<Box<Path>>,
+    paths: IndexSet<Arc<OsStr>, FxBuildHasher>,
     pub(super) linter: Linter,
     resolver: Option<Resolver>,
-    modules: ModuleCache,
+
+    #[cfg(test)]
+    pub(super) test_source: std::sync::RwLock<Option<String>>,
+}
+
+/// Output of `Runtime::process_path`
+struct ModuleProcessOutput {
+    /// All paths in `Runtime` are store as `OsStr`, because `OsStr` hash is faster
+    /// than `Path` - go checkout their source code.
+    path: Arc<OsStr>,
+    processed_module: ProcessedModule,
+}
+
+/// A module processed from a path
+#[derive(Default)]
+struct ProcessedModule {
+    /// Module records of source sections, or diagnostics if parsing failed on that section.
+    ///
+    /// Modules with special extensions such as .vue could contain multiple source sections (see `PartialLoader::PartialLoader`).
+    /// Plain ts/js modules have one section. Using `SmallVec` to avoid allocations for plain modules.
+    section_module_records: SmallVec<[Result<ResolvedModuleRecord, Vec<OxcDiagnostic>>; 1]>,
+
+    /// Source code and semantic of the module.
+    ///
+    /// This value is required for linter to run on the module. If import plugin is enabled,
+    /// dependencies are also processed as `ProcessedModule` to construct the module graph, but
+    /// not for linting. For these modules, `content` is set to `None`.
+    content: Option<ModuleContent>,
+}
+
+/// ModuleRecord with all specifiers in import statements resolved to real paths.
+struct ResolvedModuleRecord {
+    module_record: Arc<ModuleRecord>,
+    requested_module_paths: Vec<(/*specifier*/ CompactStr, Arc<OsStr>)>,
+}
+
+self_cell! {
+    struct ModuleContent {
+        owner: ModuleContentOwner,
+        #[not_covariant]
+        dependent: SectionContents,
+    }
+}
+// Safety: dependent borrows from owner. They're safe to be sent together.
+unsafe impl Send for ModuleContent {}
+
+struct ModuleContentOwner {
+    source_text: String,
+    allocator: Allocator,
+}
+
+/// source text and semantic for each source section. They are in the same order as `ProcessedModule.section_module_records`
+type SectionContents<'a> = SmallVec<[SectionContent<'a>; 1]>;
+struct SectionContent<'a> {
+    source: JavaScriptSource<'a>,
+    /// None if section parsing failed. The corresponding item with the same index in
+    /// `ProcessedModule.section_module_records` would be `Err(Vec<OxcDiagnostic>)`.
+    semantic: Option<Semantic<'a>>,
+}
+
+/// A module ready for linting. A `EntryModule` is generated for each path in `runtime.paths`
+///
+/// It's basically the same as `ProcessedModule`, except `content` is non-Option.
+struct EntryModule {
```

Comment:
`Entry` is a really vague term ... We need a longer and more descriptive name.

Reply:
Renamed to `ModuleToLint`, as opposed to modules required by the import plugin, not for linting. Let's me know if you have better ideas.